cli/cli-extensions.md

```diff
@@ -0,0 +1,195 @@
+
+And that can do anything programmatically and also get the benefit of having access to the Temporal client. So it could
+access some company resource (e.g. a database) and send an update to a workflow with some value.
+
+### Contributions
+
+Custom extensions allow an ecosystem of CLI extensions to form that do not require Temporal oversight. For example
+
+    temporal workflow show-diagram --workflow-id foo-id
+
+Could be an extension that downloads the workflow history and makes a nice visualization of it.
+
+### Cloud CLI
+
+A Temporal Cloud extension will be available (and likely released alongside the traditional CLI) that will allow
+cloud-specific operations like:
+
+    temporal cloud namespace list
+
+This can share authentication with the CLI in many ways. It can even support login, so the following two commands could
+work well together:
+
+    temporal cloud login --profile cloud
+    temporal workflow start --profile cloud --type foo --task-queue some-tq --workflow-id foo-id
+
+These are just usage examples for the purposes of explaining extensions, this may not be what `temporal cloud` extension
+looks like when built.
+
+## Runtime Behavior
+
+### Lookup
+
+Extensions are any executables on the `PATH` with `temporal-` prefix. Extension executable names use `-` for each
+subcommand and `_` for each dash in the command (but still can be called with an underscore). This is similar to
+`kubectl` behavior.
+
+The `PATH` lookup is most-specific to least-specific. For example, running `temporal foo bar-baz qux` will try to match
+the following in order:
+
+* `temporal-foo-bar_baz-qux`
+* `temporal-foo-bar_baz`
+* `temporal-foo`
+
+And the first one found will be executed with all arguments. And since underscores and dashes are the same,
+`temporal-foo-bar_baz` executable is called for both `temporal foo bar-baz qux` and `temporal foo bar_baz qux` though
+help text only shows the former.
+
+Built-in commands cannot be overridden by extensions, but subcommands can be added to existing commands.
+
+### Discovery and Help Text
+
+Running `temporal --help` (or `temporal help`) _does not_ list extensions. But running `temporal help -a` (or `--all`)
+_does_ list all extensions by traversing the `PATH` and getting every `temporal-`-prefixed executable. There is no short
+description for any of these extensions, they are simply shown as available external commands. All extensions on the
+`PATH` are shown as they are, even if they are multiple commands deep. For example `temporal-foo-bar` and
+`temporal-workflow-dosomething-somethingelse` are shown as `foo bar` and `workflow dosomething somethingelse` in the
+list respectively, not as just `foo` or implied as part of the `workflow` command.
+
+Extensions on built-in commands are not shown as part of the parent command's help.
+
+Running `temporal help <command sequence>` is treated as `temporal <command sequence> --help`. So the same style of
+lookup is performed for extensions.
```

Comment:
Having a `temporal plugin` subcommand could be worth considering for discoverability (e.g. `temporal plugin list`)

Reply:
I don't think we should call these "plugins", but regardless, I found the `git` approach a bit simpler than a full, separate command for listing.

Reply:
The advantage of a subcommand is that when a user does `temporal --help`, they'll see a subcommand with a name like "extension", alerting them to the fact that extensions are a possibilty. Not blocking, but worth considering for discoverability.

Reply:
We can add a "To see all extensions, pass --all" to the end of `--help` if it's important that default help shows that extensions are thing, but I am hoping to avoid a new sub command at least not until there is more to do that just list commands.